### PR TITLE
[RHOL-435]: Adds support for using secp256k1 crypto library

### PR DESCRIPTION
**Pull request by goral09**
## Overview
Adds support for using secp256k1 crypto library. For now build process just builds library for linux x64. Need to be enriched with build process for other operating systems.

One has to call `secp256k1/secp256k1:generate` (just as one would execute `bnfc:generate`) to build native library and get `*.so` file available in the `crypto` project.

### Does this PR relate to an RChain JIRA issue? 
[JIRA issue](https://rchain.atlassian.net/browse/RHOL-435).

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
For now build process just builds library for linux x64. Need to be enriched with build process for other operating systems.

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -51,6 +51,7 @@ object Dependencies {
   val scodecBits          = "org.scodec"                 %% "scodec-bits"               % "1.1.5"
   val shapeless           = "com.chuusai"                %% "shapeless"                 % "2.3.2"
   val weupnp              = "org.bitlet"                  % "weupnp"                    % "0.1.+"
+  val apacheLang     = "org.apache.commons" % "commons-lang3" % "3.7"
```
Comment (henrytill) [inline project/Dependencies.scala]: alignment

File: secp256k1Native/src/main/java/org/bitcoin/NativeSecp256k1.java [hunk 2]
```diff
@@ -0,0 +1,446 @@
+/*
```
Comment (henrytill) [inline secp256k1Native/src/main/java/org/bitcoin/NativeSecp256k1.java]: Are these files not part of the submodule repo? 

If you are just trying to move them into the "right" place for sbt, I would consider keeping them in their current location and changing where sbt looks for them.  This approach seems brittle in the face of upstream changes.

File: secp256k1Native/src/main/java/org/bitcoin/Secp256k1Context.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
+/*
```
Comment (henrytill) [inline secp256k1Native/src/main/java/org/bitcoin/Secp256k1Context.java]: See above.

File: .gitmodules [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[submodule "secp256k1Native/secp256k1"]
```
Comment (henrytill) [inline .gitmodules]: Oh.  Has there been a discussion about using submodules?  

I don't mind them, but I generally only use them in team settings where everybody is OK with them and are reasonably comfortable using them.

File: project/Secp256k1.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  def cleanDirs(dirs: File*): Unit = dirs.foreach(dir => s"rm -rf $dir".!)
+
+  def cleanProject(projectDir: File, dirs: File*) = {
+    cleanDirs(dirs:_*)
+    val makeDistCleanCmd = s"make -C ${projectDir.getAbsolutePath} distclean"
+    makeDistCleanCmd.!
+  }
+
+  def secp256k1Generate(baseDir: File) = {
+    val autogenCmd: String = "./autogen.sh"
```
Comment (henrytill) [inline project/Secp256k1.scala]: So now we depend on `autotools`

File: project/Secp256k1.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    val autogenProcess = Process(autogenCmd, new File(s"${baseDir.getAbsolutePath}/secp256k1/"))
+    val configureCmd: String = "./configure --enable-jni --enable-experimental --enable-module-ecdh"
+    val configureProcess = Process(configureCmd, new File(s"${baseDir.getAbsolutePath}/secp256k1/"))
+    val makeCmd: String = s"make -C ${baseDir.getAbsolutePath}/secp256k1"
```
Comment (henrytill) [inline project/Secp256k1.scala]: And `make`

File: project/Secp256k1.scala [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    val makeProcess = Process(makeCmd)
+    val binariesSourceFolder = baseDir / "secp256k1" / ".libs"
+    val resourcesFolder = baseDir / "src" / "main" / "resources"
+    val makeResourcesFolderCmd = s"mkdir ${resourcesFolder.getAbsolutePath}"
+    val mkResourcesFolderProcess = Process(makeResourcesFolderCmd)
+    val copyBinariesCmd: String = s"cp libsecp256k1.so ${resourcesFolder.getAbsolutePath}/"
```
Comment (henrytill) [inline project/Secp256k1.scala]: This doesn't work on macOS or Windows

File: secp256k1Native/src/main/java/org/bitcoin/NativeSecp256k1.java (hunk 2 above)
Comment (goral09) [reply secp256k1Native/src/main/java/org/bitcoin/NativeSecp256k1.java]: Not exactly the same as original ones, I added few lines that should take care of loading proper binaries depending on the OS.

File: .gitmodules (hunk 4 above)
Comment (goral09) [reply .gitmodules]: I thought that was the most reasonable thing to do in this case.

File: project/Secp256k1.scala (hunk 5 above)
Comment (goral09) [reply project/Secp256k1.scala]: Yeah, that's what the build process requires.

Comment (odeac) [reply project/Secp256k1.scala]: Do we need to have secp256k1 built during our build process? Can't we just add it to the prerequisites for the dev machine?

**Comment by goral09** [comment]
@odeac 
> Do we need to have secp256k1 built during our build process? Can't we just add it to the prerequisites for the dev machine?

I would prefer to have it that way but I think that general consensus was that we should ship crypto libraries on our own. Is that right @henrytill @guardbotmk3  ?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 8]
```diff
@@ -66,15 +66,16 @@ object SystemProcesses {
 
   //TODO(mateusz.gorski): we decided to look into delivering secp256k1 library (https://github.com/bitcoin-core/secp256k1)
   // as separate jar in the future
```
[the remark this answers is not in the document]
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: No, it's not. Thanks.

**Comment by goral09** [comment]
Closing in favor of [sec256k1-java library](https://github.com/rchain/secp256k1Java).

**Pull request closed without merging**